Make it possible to merge `transform` option with preset

gziolo opened the pull request:
## Summary

I discovered that `transform` option set in Jest preset is completely overridden when the same option is provided in the Jest config that extends such preset. I run into this issue when working on the preset for WordPress core: https://github.com/WordPress/gutenberg/pull/4705. It turned out that behavior differs from another option that uses `object` as data representations: `moduleNameMapper`. This PR tries to introduce consistency in the code and allow to merge values from `transform` option set in both config and preset.

### Example

`@wordpress/jest-preset-default`:
```json
{
  "transform": {
    "^.+\\.jsx?$": "babel-jest",
    "\\.pegjs$": "<rootDir>/node_modules/@wordpress/jest-preset-default/scripts/pegjs-transform.js
  } 
}
```

`jest.config.js`:
```json
{
  "preset": "@wordpress/jest-preset-default",
  "transform": {
    "\\.pegjs$": "<rootDir>/test/unit/pegjs-transform.js"
  }
}
```

Expected output:
```json
{
  "transform": {
    "^.+\\.jsx?$": "babel-jest",
    "\\.pegjs$": "<rootDir>/test/unit/pegjs-transform.js"
  }
}
```

## Test plan

This change is limited to `jest-config. This is how it was tested:

`yarn test packages/jest-config`

thymikee commented:
Can you add a test for that? Merging transforms may cause unexpected issues though, like we had with `moduleNameMapper` (hence this weird merging logic as you can see in a block above your code)

gziolo commented:
Sure thing, I wanted to have PR to reference in the PR I'm working on. I have to override this option to make it work which is pretty unexpected when using preset. Well, it might be not that confusing if some options wouldn't work this way :)

I must admit that `moduleNameMapper` is handled in a very unexpected way. Why is there needed 1st `options.moduleNameMapper` occurrence? It's always replaced by the 2nd occurrence of the same object.

I think we can also provide some helper functions to make this easier to maintain:

```js
const mergeOptionWithPreset = (options, preset, optionName) => {
  if (options[optionName] && preset[optionName]) {
    options[optionName] = Object.assign(
      {},
      preset[optionName],
      options[optionName],
    );
  }
}  
```

It can be expressed also in a more functional way if you prefer.

thymikee commented:
See https://github.com/facebook/jest/pull/3689

gziolo commented:
So it's only about the order. It makes sense then. I will follow the same pattern. It still makes sense to do some refactoring, should I include it in this PR or open another one?

thymikee commented:
This is the same logic, so feel free to refactor it as a part of this PR, it's small enough to fit in.

gziolo commented on packages/jest-config/src/__tests__/normalize.test.js:
```diff
@@ -964,7 +995,7 @@ describe('preset without setupFiles', () => {
   beforeAll(() => {
     jest.mock(
-      '/node_modules/react-native/jest-preset.json',
+      '/node_modules/react-foo/jest-preset.json',
```
It seems to be some kind of bug here. When it mocks the same file name as in the other `describe` block, it doesn't reload the config but uses what was previously set. I updated this name to ensure that `setupFiles` is configured properly, i.e. `babel-jest` is enabled which triggers prepend of `regenerator-runtime`.

thymikee replied:
We could switch to `jest.doMock` here to avoid this, since it's already placed before the require call and we don't need to hoist it.

gziolo replied:
Let me try it. This was my guess that `jest.mock` calls get hoisted and this flow doesn't work as expected.

gziolo commented on packages/jest-config/src/normalize.js:
```diff
@@ -46,6 +46,21 @@ const PRESET_NAME = 'jest-preset' + JSON_EXTENSION;
   new ValidationError(ERROR, message, DOCUMENTATION_NOTE);
 
+const mergeOptionWithPreset = (
+  options: InitialOptions,
```
I hope it is fine. This is the first time I update code that uses Flow :)

gziolo commented on packages/jest-config/src/__tests__/normalize.test.js:
```diff
@@ -922,7 +924,10 @@ describe('preset', () => {
       '/node_modules/a',
       '/node_modules/b',
-      '/node_modules/regenerator-runtime/runtime',
+    ]);
```
`babel-jest` isn't explicitly set, so it doesn't append `regenerator-runtime` anymore.

gziolo commented:
@thymikee - I refactored code, added a new test and updated existing tests to better reflect the current behavior. There are some issues with cache on CircleCi, but I don't have permissions to reset them. Let me know if there is anything else to fix or add.

thymikee commented on packages/jest-config/src/__tests__/normalize.test.js:
```diff
@@ -951,6 +956,32 @@ describe('preset', () => {
+    // Object initializer not used for properties as a workaround for
+    //  sort-keys eslint rule while specifying properties in
+    //  non-alphabetical order for a better test
+    const transform = {};
```
You can just disable the rule for this object

gziolo replied:
I shamelessly copied from the other test, but it makes sense what you said. I will update.

gziolo replied:
Updated 👍

gziolo commented:
I tried with https://github.com/facebook/jest/pull/5505/commits/757bc3b804efcd17430fe3a54d07786341e582b4 `.doMock` & `dontMock` pair, but it didn't help much. I left them to make it more obvious that they should be used only for those specific `describe` blocks, thus shouldn't be hoisted.

thymikee commented:
`normalize.test.js` passes on Appveyor, so that should be ok.

SimenB commented:
@gziolo mind updating the changelog?

gziolo commented:
Yes, I wasn’t aware of that. I will do it by Monday 👍

gziolo commented:
@SimenB done ✅

SimenB commented:
I can confirm the full test suite passes on my machine (a mac), so no issues from the skipped windows tests either

SimenB commented on packages/jest-config/src/normalize.js:
```diff
@@ -46,6 +46,21 @@ const PRESET_NAME = 'jest-preset' + JSON_EXTENSION;
+      {},
+      options[optionName],
+      preset[optionName],
+      options[optionName],
```
why do we have `options[optionName]` twice? I realise it was the same before this patch

gziolo replied:
To preserve the order of keys in the config that is overriding preset. I had the same question initially 😃 Maybe an inline comment with explanation would help?

SimenB replied:
Yeah, I think so.
